implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_state_machine.rs

```diff
@@ -0,0 +1,256 @@
+pub(super) struct CommonStateMachine {
+    pub(super) vault: Arc<dyn XXVault>,
+    pub(super) identities: Arc<Identities>,
+    pub(super) identifier: IdentityIdentifier,
+    pub(super) credentials: Vec<Credential>,
+    pub(super) trust_policy: Arc<dyn TrustPolicy>,
+    pub(super) trust_context: Option<TrustContext>,
+    their_identifier: Option<IdentityIdentifier>,
+}
+
+impl CommonStateMachine {
+    pub(super) fn new(
+        vault: Arc<dyn XXVault>,
+        identities: Arc<Identities>,
+        identifier: IdentityIdentifier,
+        credentials: Vec<Credential>,
+        trust_policy: Arc<dyn TrustPolicy>,
+        trust_context: Option<TrustContext>,
+    ) -> Self {
+        Self {
+            vault,
+            identities,
+            identifier,
+            credentials,
+            trust_policy,
+            trust_context,
+            their_identifier: None,
+        }
+    }
+
+    /// Prepare a payload containing the identity of the current party and serialize it.
+    /// That payload contains:
+    ///
+    ///  - the current identity
+    ///  - a signature of the static key used during the handshake
+    ///  - the identity credentials
+    ///
+    pub(super) async fn make_identity_payload(&self, static_key: &KeyId) -> Result<Vec<u8>> {
+        // prepare the payload that will be sent either in message 2 or message 3
+        let identity = self
+            .identities
+            .repository()
+            .get_identity(&self.identifier)
+            .await?;
+        let payload = IdentityAndCredentials {
+            identity: identity.export()?,
+            signature: self.sign_static_key(identity, static_key).await?,
+            credentials: self.credentials.clone(),
+        };
+        Ok(serde_bare::to_vec(&payload)?)
+    }
+
+    /// Verify the identity sent by the other party: the signature and the credentials must be valid
+    /// If everything is valid, store the identity identifier which will used to make the
+    /// final state machine result
+    pub(super) async fn verify_identity(
+        &mut self,
+        peer: IdentityAndCredentials,
+        peer_public_key: &PublicKey,
+    ) -> Result<()> {
+        let identity = self.decode_identity(peer.identity).await?;
+        self.verify_signature(&identity, &peer.signature, peer_public_key)
+            .await?;
+        self.verify_credentials(&identity, peer.credentials).await?;
+        self.their_identifier = Some(identity.identifier());
+        Ok(())
+    }
+
+    /// Deserialize a payload as D from a bare encoding
+    pub(super) fn deserialize<D: for<'a> Deserialize<'a>>(payload: Vec<u8>) -> Result<D> {
+        serde_bare::from_slice(payload.as_slice())
+            .map_err(|error| Error::new(Origin::Channel, Kind::Invalid, error))
+    }
+
+    /// Sign the static key used in the key exchange with the identity private key
+    async fn sign_static_key(&self, identity: Identity, key_id: &KeyId) -> Result<Signature> {
```

Comment:
I think we should pass the signature to that state machine instead of signing here.

Reply:
This code is inside the common code for both state machine. I changed the code so that the payload preparation (and signing) is triggered before we build the `Handshake` struct, right after we generate the static key.

Reply:
Have you pushed it? I don't get what you mean just from the description

Reply:
I have pushed the change now: https://github.com/build-trust/ockam/pull/4986/commits/d7bd2ea2b3a7f15f48f5ce1dbcb33f2812e44532

Reply:
I think it will be moved much further out in the future, but we should not bother with it right now